### PR TITLE
Refactoring operator_utils

### PR DESCRIPTION
**Pull request by obriente**
operator_utils has for a while been a dumping ground for any functions that we didn't know where else to put. I've pulled all of the code into separate files with two exceptions:
--- Code for loading and saving operators, which I think can stay (perhaps we should rename the file though?).
--- Code for calculating hermitian conjugates, checking if an operator is hermitian/is the identity, which I think needs a serious refactor into SymbolicOperator and PolynomialTensor.

(I had refactored out repconversions.conversions at the same time, but let's leave this till later.)

LMK what you think.

### CODE REVIEW

**Comment by ncrubin** [comment]
There is too much going on in this PR.  An easy division is the bullet points you mention in your original message.  I don't think repconversions and opconversions should be combined.  There is a natural division on types of transformations and I think making that division clear makes development easier.

**Comment by obriente** [comment]
Let me undo the conversions mixture and raise it as an issue instead; you are right to keep this separate.

**Comment by obriente** [comment]
This circular dependency issue is a huge pain - I'm also not getting these errors on my computer for some reason. @ncrubin , any idea?

**Comment by obriente** [comment]
Alright, looks like it's finally passing tests. That was an exercise in import-chasing and frustration. Sorry for the large number of changes, seems that I managed to upset a finely balanced system by splitting this file up.

**Comment by ncrubin** [comment]
@obriente can you summarize again what is in this PR?  Is it just a pruning and redistribution of operator_utils?

**Comment by obriente** [comment]
It's a redistribution - I didn't prune any code. Most of the edits are various import changes to redirect to the right place / prevent circular imports.

**Comment by ncrubin** [comment]
@obriente Cool thanks for the reorg of this big file.  I think breaking up large files like this will continue to help skirt circular import issues.  

For my own edification on what you are doing in this PR I'm documenting here what I see are the most notable changes.

* normal_ordered is now in transforms
* functionals module is added. Note to all: Functionals take an objects and return numbers.
* contextuality testing is moving into functionals from operator_utils
* inverse_fourier_transform is now in repconversions. 
* jellium_hf_state was in utils and is now in hamiltonians (At first glance I am biased towards putting this in linalg since it produces a vector.  Is the motivation for putting it in hamiltonians because it ingests a hamiltonian and thus can be thought of as a hamiltonian utility? I've noted this above.)
* eigenspectrum is now correctly placed in linalg! Yay!
* check_no_sympy is no longer hidden and moved from repconversions to opconversions. Is the motivation for the move that the check involves Sympy ops of SymbolicOps thus opconversions is a better location for the utility?
* group_tensor_product_basis from operator_utils goes into qubit_partitioning of measurements.
* Moved reorder to opconversions because it changes SymbolicOps objects. 
* inline_sum is moved to where it is used (bravyi_kitaev)
* chemist_order moved out of operator_utils 
* freeze_orbitals into tapering off qubits (only place it is used)

Did I miss something big?

File: src/openfermion/hamiltonians/jellium_hf_state_test.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 from openfermion.linalg import (get_sparse_operator, get_ground_state,
                                 expectation, jw_number_restrict_operator)
 from openfermion.utils.grid import Grid
-from openfermion.utils.jellium_hf_state import hartree_fock_state_jellium
+from openfermion.hamiltonians.jellium_hf_state import hartree_fock_state_jellium
```
[the remark this answers is not in the document]
Comment (obriente) [reply src/openfermion/hamiltonians/jellium_hf_state_test.py:11]: I actually made this choice to avoid some circular import issues (which are mostly coming from the remaining functions in operator_utils.py), but I think I agree this eventually belongs in linalg. I think the current dependency on count_qubits will cause problems though - is it ok if we skip this discussion till I've shifted this out of operator_utils alongside the hermitian conjugation functionality?

File: src/openfermion/ops/representations/quadratic_hamiltonian.py [hunk 2]
```diff
@@ -375,6 +371,9 @@ def diagonalizing_circuit(self):
                 of modes :math:`i` and :math:`j` by angles :math:`\theta`
                 and :math:`\varphi`.
         """
+        # Adding inline import here to prevent circular issues
+        from openfermion.linalg.givens_rotations import (
```
[the remark this answers is not in the document]
Comment (obriente) [reply src/openfermion/ops/representations/quadratic_hamiltonian.py:375]: Yeah, I spent a few hours trying to hack this around and this was the only way that seemed to work in both versions. I don't like this solution, but I think the work to fix it properly is bigger than just this PR. I've opened #631 on this issue; is it sufficient if I mark this in the code as a TODO so that we remember it (as you suggested below)?

File: src/openfermion/hamiltonians/jellium_hf_state_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (obriente) [reply src/openfermion/hamiltonians/jellium_hf_state_test.py:11]: Following your suggestion below - maybe I can add a TODO to the hartree_fock_state_jellium file so that we remember that this should be shifted.

File: src/openfermion/ops/representations/quadratic_hamiltonian.py [hunk 3]
```diff
@@ -473,6 +472,7 @@ def antisymmetric_canonical_form(antisymmetric_matrix):
         canonical(ndarray): The canonical form C of antisymmetric_matrix
         orthogonal(ndarray): The orthogonal transformation R.
     """
+    from openfermion.linalg.givens_rotations import swap_columns, swap_rows
```
[the remark this answers is not in the document]
Comment (obriente) [reply src/openfermion/ops/representations/quadratic_hamiltonian.py:475]: Great idea with the TODO, thanks for the suggestion; I'll implement that.

**Comment by obriente** [comment]
I think this is it. I think check_no_sympy might have been moved as part of my initial compression of the two conversions.py files - I don't think it particularly matters where it goes (from memory both use it?)

Thanks for taking the time to compile that summary.
